chore(linter): Add a convenience script for printing a table of rule implementation stats

connorshea opened the pull request:
AI Disclosure: Written with help from GitHub Copilot + Raptor mini, reviewed by me and confirmed to be functional and correct.

 Note that the reason for the # of implemented rules being higher than the actual # of rules listed on the rules page of the docs is due to jest-vitest rules, where we implement two rules as one. (there are also a few other rules that get counted this way, not just the jest-vitest rules).

This adds a script for listing the full progress report stats across all plugins/rules. This is not used for anything yet, but could be used in the future on the website, or just as an internal tracker for our own usage.

Command to install the dependencies/latest eslint plugins and then run the script: `cd tasks/lint_rules && npm install && cd ../.. && node tasks/lint_rules/src/count-implemented.mjs`

Output:

```
eslint       Implemented:  165 /  196 ( 84.18%), Not Implemented:  31, Not Supported:  96
typescript   Implemented:  102 /  130 ( 78.46%), Not Implemented:  28, Not Supported:   3
n            Implemented:    4 /   36 ( 11.11%), Not Implemented:  32, Not Supported:   5
unicorn      Implemented:  122 /  141 ( 86.52%), Not Implemented:  19, Not Supported:   3
jsdoc        Implemented:   18 /   75 ( 24.00%), Not Implemented:  57, Not Supported:   0
import       Implemented:   30 /   44 ( 68.18%), Not Implemented:  14, Not Supported:   2
jsx-a11y     Implemented:   30 /   36 ( 83.33%), Not Implemented:   6, Not Supported:   3
jest         Implemented:   52 /   68 ( 76.47%), Not Implemented:  16, Not Supported:   3
react        Implemented:   47 /   87 ( 54.02%), Not Implemented:  40, Not Supported:  45
react-perf   Implemented:    4 /    4 (100.00%), Not Implemented:   0, Not Supported:   0
nextjs       Implemented:   21 /   21 (100.00%), Not Implemented:   0, Not Supported:   0
promise      Implemented:   16 /   16 (100.00%), Not Implemented:   0, Not Supported:   1
vitest       Implemented:   49 /   78 ( 62.82%), Not Implemented:  29, Not Supported:   2
vue          Implemented:   14 /   83 ( 16.87%), Not Implemented:  69, Not Supported: 165
-----------------------------------------------------------------------------------------
total        Implemented:  674 / 1015 ( 66.40%), Not Implemented: 341, Not Supported: 328
```

This also fixes a bug (arguably not a bug I guess? 🤷) where the script to generate the rules lists would print a bunch of irrelevant messages because we had implemented type-checked rules (which are not included in the plugin issue in oxlint, as they are for tsgolint). So we'll no longer see a bunch of lines like this in the logs:

```
👀 typescript/no-for-in-array is implemented but not found in their rules
👀 typescript/no-implied-eval is implemented but not found in their rules
👀 typescript/no-meaningless-void-operator is implemented but not found in their rules
👀 typescript/no-misused-spread is implemented but not found in their rules
👀 typescript/no-mixed-enums is implemented but not found in their rules
👀 typescript/no-redundant-type-constituents is implemented but not found in their rules
👀 typescript/no-unnecessary-boolean-literal-compare is implemented but not found in their rules
👀 typescript/no-unnecessary-template-expression is implemented but not found in their rules
👀 typescript/no-unnecessary-type-arguments is implemented but not found in their rules
👀 typescript/no-unnecessary-type-assertion is implemented but not found in their rules
```

connorshea commented on tasks/lint_rules/src/count-implemented.mjs:
```diff
@@ -0,0 +1,80 @@
+  updatePendingFixStatus,
+} from "./oxlint-rules.mjs";
+
+void (async () => {
```
That one is on me for not being familiar with top-level awaits and assuming the slop was correct 😅

camc314 commented:
### Merge activity

* **Jan 1, 11:48 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Jan 1, 11:48 AM UTC**: `camc314` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).

connorshea committed:
chore(linter): Add a convenience script for printing a table of rule implementation stats (#17543)

AI Disclosure: Written with help from GitHub Copilot + Raptor mini, reviewed by me and confirmed to be functional and correct.

 Note that the reason for the # of implemented rules being higher than the actual # of rules listed on the rules page of the docs is due to jest-vitest rules, where we implement two rules as one. (there are also a few other rules that get counted this way, not just the jest-vitest rules).

This adds a script for listing the full progress report stats across all plugins/rules. This is not used for anything yet, but could be used in the future on the website, or just as an internal tracker for our own usage.

Command to install the dependencies/latest eslint plugins and then run the script: `cd tasks/lint_rules && npm install && cd ../.. && node tasks/lint_rules/src/count-implemented.mjs`

Output:

```
eslint       Implemented:  165 /  196 ( 84.18%), Not Implemented:  31, Not Supported:  96
typescript   Implemented:  102 /  130 ( 78.46%), Not Implemented:  28, Not Supported:   3
n            Implemented:    4 /   36 ( 11.11%), Not Implemented:  32, Not Supported:   5
unicorn      Implemented:  122 /  141 ( 86.52%), Not Implemented:  19, Not Supported:   3
jsdoc        Implemented:   18 /   75 ( 24.00%), Not Implemented:  57, Not Supported:   0
import       Implemented:   30 /   44 ( 68.18%), Not Implemented:  14, Not Supported:   2
jsx-a11y     Implemented:   30 /   36 ( 83.33%), Not Implemented:   6, Not Supported:   3
jest         Implemented:   52 /   68 ( 76.47%), Not Implemented:  16, Not Supported:   3
react        Implemented:   47 /   87 ( 54.02%), Not Implemented:  40, Not Supported:  45
react-perf   Implemented:    4 /    4 (100.00%), Not Implemented:   0, Not Supported:   0
nextjs       Implemented:   21 /   21 (100.00%), Not Implemented:   0, Not Supported:   0
promise      Implemented:   16 /   16 (100.00%), Not Implemented:   0, Not Supported:   1
vitest       Implemented:   49 /   78 ( 62.82%), Not Implemented:  29, Not Supported:   2
vue          Implemented:   14 /   83 ( 16.87%), Not Implemented:  69, Not Supported: 165
-----------------------------------------------------------------------------------------
total        Implemented:  674 / 1015 ( 66.40%), Not Implemented: 341, Not Supported: 328
```

This also fixes a bug (arguably not a bug I guess? 🤷) where the script to generate the rules lists would print a bunch of irrelevant messages because we had implemented type-checked rules (which are not included in the plugin issue in oxlint, as they are for tsgolint). So we'll no longer see a bunch of lines like this in the logs:

```
👀 typescript/no-for-in-array is implemented but not found in their rules
👀 typescript/no-implied-eval is implemented but not found in their rules
👀 typescript/no-meaningless-void-operator is implemented but not found in their rules
👀 typescript/no-misused-spread is implemented but not found in their rules
👀 typescript/no-mixed-enums is implemented but not found in their rules
👀 typescript/no-redundant-type-constituents is implemented but not found in their rules
👀 typescript/no-unnecessary-boolean-literal-compare is implemented but not found in their rules
👀 typescript/no-unnecessary-template-expression is implemented but not found in their rules
👀 typescript/no-unnecessary-type-arguments is implemented but not found in their rules
👀 typescript/no-unnecessary-type-assertion is implemented but not found in their rules
```
